go/adbc/driver/bigquery/connection.go

```diff
@@ -0,0 +1,868 @@
+	if err != nil {
+		return 0, 0, adbc.Error{
+			Code: adbc.StatusInvalidData,
+			Msg:  fmt.Sprintf("Cannot parse scale `%s`: %s", values[1], err.Error()),
+		}
+	}
+	return int32(precision), int32(scale), nil
+}
+
+func patternToRegexp(pattern *string) (*regexp.Regexp, error) {
+	patternString := ""
+	if pattern != nil {
+		patternString = *pattern
+	}
+	patternString = strings.TrimSpace(patternString)
+
+	convertedPattern := ".*"
+	if patternString != "" {
+		convertedPattern = fmt.Sprintf("(?i)^%s$", strings.ReplaceAll(strings.ReplaceAll(patternString, "_", "."), "%", ".*"))
+	}
+	r, err := regexp.Compile(convertedPattern)
+	if err != nil {
+		return nil, adbc.Error{
+			Code: adbc.StatusInvalidArgument,
+			Msg:  fmt.Sprintf("Cannot parse pattern `%s`: %s", patternString, err.Error()),
+		}
+	}
+	return r, nil
+}
+
+func (c *connectionImpl) getDatasetsInProject(ctx context.Context, projectIDPattern, datasetsIDPattern *string, cb FilteredDatasetHandler) error {
+	pattern, err := patternToRegexp(projectIDPattern)
+	if err != nil {
+		return err
+	}
+	if !pattern.MatchString(c.client.Project()) {
+		return iterator.Done
+	}
+
+	pattern, err = patternToRegexp(datasetsIDPattern)
+	if err != nil {
+		return err
+	}
+
+	it := c.client.Datasets(ctx)
+	for {
+		dataset, err := it.Next()
+		if err != nil {
+			if errors.Is(err, iterator.Done) {
+				break
+			}
+			return err
+		}
+
+		if pattern.MatchString(dataset.DatasetID) {
+			err = cb(dataset)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func getTablesInDataset(ctx context.Context, dataset *bigquery.Dataset, tableNamePattern *string, cb FilteredTableHandler) error {
+	pattern, err := patternToRegexp(tableNamePattern)
+	if err != nil {
+		return err
+	}
+
+	it := dataset.Tables(ctx)
+	for {
+		table, err := it.Next()
+		if err != nil {
+			if errors.Is(err, iterator.Done) {
+				break
+			}
+			return err
+		}
+
+		if pattern.MatchString(table.TableID) {
+			err = cb(table)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (c *connectionImpl) getTableSchemaWithFilter(ctx context.Context, catalog *string, dbSchema *string, tableName string, columnName *string) (*arrow.Schema, error) {
+	if catalog == nil {
+		catalog = &c.catalog
+	}
+	sanitizedCatalog, err := sanitize(*catalog)
+	if err != nil {
+		return nil, err
+	}
+
+	if dbSchema == nil {
+		dbSchema = &c.dbSchema
+	}
+	sanitizedDbSchema, err := sanitize(*dbSchema)
+	if err != nil {
+		return nil, err
+	}
+
+	pattern, err := patternToRegexp(columnName)
+	if err != nil {
+		return nil, err
+	}
+
+	// sadly that query parameters cannot be used in place of table names
+	queryString := fmt.Sprintf("SELECT * FROM `%s`.`%s`.INFORMATION_SCHEMA.COLUMNS WHERE table_name = @tableName", sanitizedCatalog, sanitizedDbSchema)
+	query := c.client.Query(queryString)
+	query.Parameters = []bigquery.QueryParameter{
+		{
+			Name:  "tableName",
+			Value: tableName,
+		},
+	}
```

Comment:
can't we do this for the other cases (datasets and projects) too? rather than having to manually filter/process them?

Reply:
I'm so sorry I might be completely wrong with the following SQL query but it seems that BigQuery doesn't support this.

<img width="611" alt="Screenshot 2024-04-25 at 05 38 06" src="https://github.com/apache/arrow-adbc/assets/89497197/6cfb5454-ddd7-4fba-a561-90cdd2e65850">

Reply:
As for using `INFORMATION_SCHEMA.SCHEMATA` it should return these information per [the docs here](https://cloud.google.com/bigquery/docs/information-schema-datasets-schemata) but I had no luck with it either using API or in their cloud console.

I'm not quite sure about the reason, maybe there're some options somewhere I have to set and they didn't mention these option (that might be totally apparent to SQL experts) in the docs.

<img width="591" alt="Screenshot 2024-04-25 at 05 45 55" src="https://github.com/apache/arrow-adbc/assets/89497197/ab437e6f-ab4b-4fb3-bb25-09be5b2245df">
<img width="598" alt="Screenshot 2024-04-25 at 05 48 36" src="https://github.com/apache/arrow-adbc/assets/89497197/97c30a09-5aef-4244-88a9-6662ad04571a">